### PR TITLE
Avoid expensive tracking objects for prepared statements

### PR DESCRIPTION
**Pull request by SamSaffron**
Per #36949 we introduce a race condition fix for #36763

This refines the fix to avoid using `Concurrent::ThreadLocalVar`

The implementation in the concurrent lib is rather expensive, culminating in
a finalizer per object that spins off a thread to do cleanup work.

None of this expense is needed as we can simply implement
the desired behavior using Ruby primitives. Additionally this moves to a Fiber bound implementation vs a thread bound implementation, something that is not desired for this particular usage. 

cc @matthewd / @rafaelfranca / @97jaz

### CODE REVIEW

**Comment by kaspth** [comment]
Are we sure we need the extra instance variable versus something in this direction?

```ruby
module PreparedStatementsRegistry
  extend ActiveSupport::PerThreadRegistry

  attr_accessor :enabled

  def disable
    old_enabled, self.enabled = self.enabled, false
    yield
  ensure
    self.enabled = old_enabled
  end
end

if self.class.type_cast_config_to_boolean(config.fetch(:prepared_statements) { true })
  PreparedStatementsRegistry.enabled = true
  @visitor.extend(DetermineIfPreparableVisitor)
else
  PreparedStatementsRegistry.enabled = true
end

def prepared_statements
  PreparedStatementsRegistry.enabled
end

def unprepared_statement(&block)
  PreparedStatementsRegistry.disable(&block)
end
```

**Comment by SamSaffron** [comment]
An interesting minor advantage of the proposal I made is that it is pretty much free if you are not using prepared statements, we don't even store anything in the thread local storage. 

I am not sure per thread registry will work cause it is based off an object name per https://github.com/SamSaffron/rails/blob/f9e843b25efc35ecb47dc1bb4b2aac2fc7239c88/activesupport/lib/active_support/per_thread_registry.rb#L43-L43

So I am guessing what will happen is when to connections are on the same thread and control is yielded between 2 fibers, weird can happen. 

Honestly I am not sure how people feel about allowing the same connection to be used at the same time from 2 threads, this is not something we ever practice at Discourse. The whole underlying premise of the original change is somewhat confusing to me.

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb [hunk 1]
```diff
@@ -226,7 +238,17 @@ def seconds_idle # :nodoc:
       end
 
       def unprepared_statement
-        @prepared_statement_status.bind(false) { yield }
+        hash = nil
+
+        if @prepared_statements_default == false
```
Comment (gklsan) [inline activerecord/lib/active_record/connection_adapters/abstract_adapter.rb]: Why not we go like this?
```suggestion
        if @prepared_statements_default
          hash = Thread.current[:ar_prepared_statements] ||= {}
          hash[self.object_id] = false
        end
```

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb [hunk 2]
```diff
@@ -138,8 +137,21 @@ def schema_migration # :nodoc:
...
+        @prepared_statement_state_key ||= "_ar_prep_statements_#{SecureRandom.hex}"
+      end
+
       def prepared_statements
-        @prepared_statement_status.value
+        if @prepared_statements_default == false
+          false
+        else
+          hash = Thread.current[:ar_prepared_statements]
+          if hash && hash[self.object_id] == false
```
Comment (gklsan) [inline activerecord/lib/active_record/connection_adapters/abstract_adapter.rb]:
```suggestion
          hash && !hash[self.object_id] ? false : true
```

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb (hunk 1 above)
Comment (SamSaffron) [reply activerecord/lib/active_record/connection_adapters/abstract_adapter.rb]: Thanks, cleaned this up!

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb (hunk 2 above)
Comment (SamSaffron) [reply activerecord/lib/active_record/connection_adapters/abstract_adapter.rb]: Thanks, cleaned this up.

**Comment by kaspth** [comment]
@SamSaffron good points! I revised the patch a little further and merged it in b9cc147241. Thanks so much! 😊

**Comment by SamSaffron** [comment]
@kaspth thanks heaps, nice improvement using a Set there, it is a far better fit. 

Any chance you can backport this to the next 6 release?

**Comment by rafaelfranca** [comment]
Backported in 46cc85d534]

**Comment by kaspth** [comment]
Thanks, I forgot to backport it ✌️

**Pull request closed without merging**